### PR TITLE
test: add unit tests for core/utils and improve nil-safety in helpers

### PR DESCRIPTION
**Pull request by kincoy**
#### What type of PR is this?

/kind cleanup

---

#### What this PR does / why we need it:

This PR improves unit test coverage for several utility functions in `core/utils`, helping ensure correctness and robustness of logic that supports autoscaling decisions.

Specifically:
- Adds unit tests for:
  - `FilterOutNodesFromNotAutoscaledGroups`
  - `GetOldestCreateTimeWithGpu`
  - `hasHardInterPodAffinity`
  - `isVirtualNode`
- Fixes potential panic in `isVirtualNode` when passed a `nil` node
- Brings test coverage in `core/utils/utils.go` **from **47.4% to 84.2%****

before:
image

after:
image

---

#### Which issue(s) this PR fixes:

N/A — test coverage improvement

---

#### Special notes for your reviewer:

- Safe refactor: no production logic changed, only a nil check added in `isVirtualNode`
- Tests are table-driven and use testify assertions

---

#### Does this PR introduce a user-facing change?

```release-note
NONE

### CODE REVIEW

**Comment by kincoy** [comment]
@feiskyer @vadasambar Hi, could you please re-run the failed workflows? They appear to be unrelated to this PR. Thanks🙏!

**Comment by vadasambar** [comment]
@kincoy sorry I don't have permissions to restart the failed workflow. Can you try creating an empty commit like this to trigger the build (another option is to create a small harmless change like adding new line and removing it later):
```
git commit --allow-empty -m "Trigger Build"
```

**Comment by kincoy** [comment]
> @kincoy sorry I don't have permissions to restart the failed workflow. Can you try creating an empty commit like this to trigger the build (another option is to create a small harmless change like adding new line and removing it later):
> 
> ```
> git commit --allow-empty -m "Trigger Build"
> ```

Thanks for your feedback. I found a minor issue in my changes, which has now been fixed. I’ve pushed the update and all tests have passed. Please help review the latest commit to see if it looks good. Thanks again!🙏。

File: cluster-autoscaler/core/utils/utils.go [hunk 1]
```diff
@@ -28,9 +28,17 @@ import (
 	"k8s.io/autoscaler/cluster-autoscaler/utils/gpu"
 )
 
+const (
+	// VirtualNodeLabelValue is the value of the label that is set on virtual kubelet
+	VirtualNodeLabelValue = "virtual-kubelet"
```
Comment (jackfrancis) [inline cluster-autoscaler/core/utils/utils.go]: + 1 on the const, let's use `VirtualKubeletNodeLabelValue`

File: cluster-autoscaler/core/utils/utils.go [hunk 2]
```diff
@@ -28,9 +28,17 @@ import (
...
+)
+
 // isVirtualNode determines if the node is created by virtual kubelet
 func isVirtualNode(node *apiv1.Node) bool {
-	return node.ObjectMeta.Labels["type"] == "virtual-kubelet"
+	if node == nil {
```
Comment (jackfrancis) [inline cluster-autoscaler/core/utils/utils.go:38]: `return node != nil && node.ObjectMeta.Labels["type"] == VirtualKubeletNodeLabelValue` is more concise

File: cluster-autoscaler/core/utils/utils.go [hunk 3]
```diff
@@ -28,9 +28,17 @@ import (
 	"k8s.io/autoscaler/cluster-autoscaler/utils/gpu"
 )
 
+const (
+	// VirtualNodeLabelValue is the value of the label that is set on virtual kubelet
+	VirtualNodeLabelValue = "virtual-kubelet"
+)
+
 // isVirtualNode determines if the node is created by virtual kubelet
 func isVirtualNode(node *apiv1.Node) bool {
```
Comment (jackfrancis) [inline cluster-autoscaler/core/utils/utils.go]: while we're here, because we're only using this func specifically to detect virtual-kubelet nodes, let's rename this func to `isVirtualKubeletNode` to be more clear (and update all usages)

File: cluster-autoscaler/core/utils/utils.go (hunk 2 above)
Comment (kincoy) [reply cluster-autoscaler/core/utils/utils.go:38]: Thanks for the suggestion! I agree the one-liner is more concise, but I opted for the explicit `if node == nil` form to improve readability and maintainability.

It's quite common to use explicit nil checks, especially when dealing with nested struct access, as it allows easier debugging and extension in the future (e.g., adding logs or metrics inside the `nil` branch).

Happy to change if we prefer the concise form globally.

File: cluster-autoscaler/core/utils/utils.go (hunk 1 above)
Comment (kincoy) [reply cluster-autoscaler/core/utils/utils.go]: Updated as suggested, thank you!

File: cluster-autoscaler/core/utils/utils.go (hunk 3 above)
Comment (kincoy) [reply cluster-autoscaler/core/utils/utils.go]: Updated as suggested, thank you!

**Comment by kincoy** [comment]
Friendly ping @jackfrancis — this PR has been idle for a while. Would appreciate a review when convenient 🙏

**Review by jackfrancis — APPROVED** [review]
/lgtm
/approve

thanks for your patience and contribution @kincoy!
